File: Data-Agent-Server/data-agent-server-app/src/main/java/edu/zsc/ai/service/DriverService.java
package edu.zsc.ai.service;

import edu.zsc.ai.model.dto.response.db.AvailableDriverResponse;
import edu.zsc.ai.model.dto.response.db.InstalledDriverResponse;

import java.nio.file.Path;
import java.util.List;

/**
 * Service interface for managing JDBC driver downloads and installations.
 */
public interface DriverService {
    
    /**
     * Download a driver from Maven Central.
     *
     * @param databaseType database type (e.g., "MySQL")
     * @param version driver version (optional, uses latest if null)
     * @return path to downloaded driver file
     */
    Path downloadDriver(String databaseType, String version);
    
    /**
     * List all available driver versions from Maven Central.
     * Queries the remote Maven repository to show all downloadable versions.
     *
     * @param databaseType database type (required)
     * @return list of available driver versions from Maven Central with installation status
     */
    List<AvailableDriverResponse> listAvailableDrivers(String databaseType);
    
    /**
     * List locally installed/downloaded drivers.
     * Scans the local drivers directory.
     *
     * @param databaseType database type (required)
     * @return list of installed drivers on local disk
     */
    List<InstalledDriverResponse> listInstalledDrivers(String databaseType);
    
    /**
     * Delete a locally installed driver.
     *
     * @param databaseType database type (e.g., "MySQL")
     * @param version driver version
     */
    void deleteDriver(String databaseType, String version);
}

File: Data-Agent-Client/src/services/driver.service.ts
import http from '../lib/http';
import type {
  AvailableDriverResponse,
  InstalledDriverResponse,
  DownloadDriverResponse,
} from '../types/driver';

export const driverService = {
  /**
   * List driver versions available from Maven Central for the given database type.
   */
  listAvailableDrivers: async (databaseType: string): Promise<AvailableDriverResponse[]> => {
    const response = await http.get<AvailableDriverResponse[]>('/drivers/available', {
      params: { databaseType },
    });
    return response.data;
  },

  /**
   * List locally installed drivers for the given database type.
   */
  listInstalledDrivers: async (databaseType: string): Promise<InstalledDriverResponse[]> => {
    const response = await http.get<InstalledDriverResponse[]>('/drivers/installed', {
      params: { databaseType },
    });
    return response.data;
  },

  /**
   * Download a driver from Maven Central. If version is omitted, latest is used.
   */
  downloadDriver: async (
    databaseType: string,
    version?: string
  ): Promise<DownloadDriverResponse> => {
    const response = await http.post<DownloadDriverResponse>('/drivers/download', {
      databaseType,
      version,
    });
    return response.data;
  },

  /**
   * Delete a locally installed driver.
   */
  deleteDriver: async (databaseType: string, version: string): Promise<void> => {
    await http.delete(`/drivers/${encodeURIComponent(databaseType)}/${encodeURIComponent(version)}`);
  },
};
